File: Examples/CSharp/x-IMU3-Examples/Examples/MuxScanner.cs
namespace Ximu3Examples
{
    class MuxScanner
    {
        public MuxScanner()
        {
            // Search for connection
            var usbDevices = Ximu3.PortScanner.ScanFilter(Ximu3.CApi.XIMU3_PortType.XIMU3_PortTypeUsb);

            System.Threading.Thread.Sleep(1000); // wait for OS to release port

            if (usbDevices.Length == 0)
            {
                Console.WriteLine("No USB connections available");
                return;
            }

            Console.WriteLine("Found " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_device_to_string(usbDevices[0])));

            // Open connection
            var usbConnection = new Ximu3.Connection(Ximu3.ConnectionConfig.From(usbDevices[0])!);

            var result = usbConnection.Open();

            if (result != Ximu3.CApi.XIMU3_Result.XIMU3_ResultOk)
            {
                Console.WriteLine("Unable to open " + usbConnection.GetConfig() + ". " + Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_result_to_string(result)) + ".");
                return;
            }

            // Mux scanner (blocking)
            var muxDevices = Ximu3.MuxScanner.Scan(usbConnection);

            PrintDevices(muxDevices);

            // Mux scanner (non-blocking)
            var muxScanner = new Ximu3.MuxScanner(usbConnection, Callback);

            System.Threading.Thread.Sleep(60000);

            // Close connection
            usbConnection.Close();
        }

        private void Callback(Ximu3.CApi.XIMU3_Device[] devices)
        {
            PrintDevices(devices);
        }

        private static void PrintDevices(Ximu3.CApi.XIMU3_Device[] devices)
        {
            Console.WriteLine(devices.Length + " device(s) found");

            foreach (var device in devices)
            {
                Console.WriteLine(
                    Ximu3.Helpers.ToString(device.device_name) + ", " +
                    Ximu3.Helpers.ToString(device.serial_number) + ", " +
                    Ximu3.ConnectionConfig.From(device)
                );
                // Console.WriteLine(Ximu3.Helpers.ToString(Ximu3.CApi.XIMU3_device_to_string(device))); // alternative to above
            }
        }
    }
}

File: Examples/Python/mux_scanner.py
import time

import ximu3

# Search for connection
usb_devices = ximu3.PortScanner.scan_filter(ximu3.PORT_TYPE_USB)

time.sleep(1)  # wait for OS to release port

if not usb_devices:
    raise Exception("No USB connections available")

print(f"Found {usb_devices[0]}")

# Open connection
usb_connection = ximu3.Connection(usb_devices[0].connection_config).open()


# Mux scanner (blocking)
def print_devices(devices: list[ximu3.Device]) -> None:
    print(f"{len(devices)} device(s) found")

    for device in devices:
        print(", ".join([device.device_name, device.serial_number, str(device.connection_config)]))
        # print(device)  # alternative to above


mux_devices = ximu3.MuxScanner.scan(usb_connection)

print_devices(mux_devices)


# Mux scanner (non-blocking)
def callback(mux_devices: list[ximu3.Device]) -> None:
    print_devices(mux_devices)


_ = ximu3.MuxScanner(usb_connection, callback)

time.sleep(60)

# Close connection
usb_connection.close()
